rasa/nlu/featurizers/sparse_featurizer/lexical_syntactic_featurizer.py

```diff
@@ -189,14 +189,34 @@ def _tokens_to_features(self, tokens: List[Token]) -> List[Dict[Text, Any]]:
         configured_features = self.component_config["features"]
         sentence_features = []
 
-        for token_idx in range(len(tokens)):
+        # A function can be applied to the same token multiple times if
+        # if configured_features includes functions of the window around a token.
+        # This can become unnecessarily slow if the window size is large.
+        # For efficiency, we memoize the function call on each token.
+        functions = set(
+            [
+                function
+                for token_functions in configured_features
+                for function in token_functions
+            ]
+        )
+        token_to_function_values = []
+        for token_idx, token in enumerate(tokens):
+            function_values = {}
+            for function in functions:
+                function_values[function] = self._get_feature_value(
+                    function, token, token_idx, len(tokens)
+                )
+            token_to_function_values.append(function_values)
```

Comment:
I think it would also help to move this memoization bit to an independent function. Would be easier to follow IMO